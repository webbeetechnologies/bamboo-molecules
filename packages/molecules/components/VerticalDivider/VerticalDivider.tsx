import { memo, useMemo } from 'react';
import { type ViewStyle, type StyleProp, type ViewProps, View } from 'react-native';
import { StyleSheet } from 'react-native-unistyles';

import { registerComponentStyles, getRegisteredMoleculesComponentStyles } from '../../core';

export type Props = Omit<ViewProps, 'children'> & {
    /**
     *  Top Inset of the Divider.
     */
    topInset?: number;
    /**
     *  Bottom Inset of the Divider.
     */
    bottomInset?: number;
    /**
     *  Whether divider should be bolded.
     */
    bold?: boolean;
    /**
     *  Horizontal spacing of the Divider
     */
    spacing?: number;
    style?: StyleProp<ViewStyle>;
};

/**
 * A divider is a thin, lightweight separator that groups content in lists and page layouts.
 *
 * <div class="screenshots">
 *  <figure>
 *    <img class="medium" src="screenshots/divider.png" />
 *  </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { View } from 'react-native';
 * import { Divider, Text } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <View>
 *     <Text>Lemon</Text>
 *     <Divider />
 *     <Text>Mango</Text>
 *     <Divider />
 *   </View>
 * );
 *
 * export default MyComponent;
 * ```
 */

const VerticalDivider = ({
    topInset = 0,
    bottomInset = 0,
    spacing = 0,
    style,
    bold = false,
    ...rest
}: Props) => {
    verticalDividerStyles.useVariants({
        isBold: bold,
    });

    const memoizedStyles = useMemo(() => {
        return [
            verticalDividerStyles.root,
            style,
            topInset && { marginTop: topInset },
            bottomInset && { marginBottom: bottomInset },
            spacing && { marginHorizontal: spacing },
        ] as StyleProp<ViewStyle>;
    }, [bottomInset, style, spacing, topInset]);

    return <View {...rest} style={memoizedStyles} />;
};

export const verticalDividerStylesDefault = StyleSheet.create(theme => ({
    root: {
        width: StyleSheet.hairlineWidth,
        background: theme.colors.outlineVariant,

        variants: {
            isBold: {
                true: {
                    width: 1,
                },
            },
        },
    },
}));

registerComponentStyles('VerticalDivider', verticalDividerStylesDefault);
export const verticalDividerStyles = getRegisteredMoleculesComponentStyles('HorizontalDivider');

export default memo(VerticalDivider);
