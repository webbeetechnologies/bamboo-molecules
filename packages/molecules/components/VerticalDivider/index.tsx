import { getRegisteredComponentWithFallback, registerMoleculesComponents } from '../../core';
import VerticalDividerDefault from './VerticalDivider';

registerMoleculesComponents({
    VerticalDivider: VerticalDividerDefault,
});

export const VerticalDivider = getRegisteredComponentWithFallback(
    'VerticalDivider',
    VerticalDividerDefault,
);

export { Props as VerticalDividerProps, verticalDividerStyles } from './VerticalDivider';
