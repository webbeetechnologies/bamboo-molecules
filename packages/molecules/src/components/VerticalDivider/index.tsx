import { getRegisteredMoleculesComponent, registerMoleculesComponents } from '../../core';
import VerticalDividerDefault from './VerticalDivider';

registerMoleculesComponents({
    VerticalDivider: VerticalDividerDefault,
});

export const VerticalDivider = getRegisteredMoleculesComponent('VerticalDivider');

export { Props as VerticalDividerProps, verticalDividerStyles } from './VerticalDivider';
