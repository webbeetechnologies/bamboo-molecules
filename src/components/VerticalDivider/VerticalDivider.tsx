import { memo } from 'react';
import { StyleSheet, ViewStyle, StyleProp } from 'react-native';
import type { ViewProps } from '@webbee/bamboo-atoms';
import { useComponentTheme, useMolecules } from '../../hooks';
import { withNormalizedStyleProp } from '../../hocs';

export type Props = Omit<ViewProps, 'children'> & {
    /**
     *  Top Inset of the Divider.
     */
    topInset?: number;
    /**
     *  Bottom Inset of the Divider.
     */
    bottomInset?: number;
    /**
     *  Whether divider should be bolded.
     */
    bold?: boolean;
    /**
     *  Horizontal spacing of the Divider
     */
    spacing?: number;
    style?: StyleProp<ViewStyle>;
};

/**
 * A divider is a thin, lightweight separator that groups content in lists and page layouts.
 *
 * <div class="screenshots">
 *  <figure>
 *    <img class="medium" src="screenshots/divider.png" />
 *  </figure>
 * </div>
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { View } from 'react-native';
 * import { Divider, Text } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <View>
 *     <Text>Lemon</Text>
 *     <Divider />
 *     <Text>Mango</Text>
 *     <Divider />
 *   </View>
 * );
 *
 * export default MyComponent;
 * ```
 */

const VerticalDivider = ({
    topInset = 0,
    bottomInset = 0,
    spacing = 0,
    style,
    bold = false,
    ...rest
}: Props) => {
    const { View } = useMolecules();
    const { bold: boldStyles, ...defaultStyles } = useComponentTheme('VerticalDivider');

    return (
        <View
            {...rest}
            style={[
                defaultStyles,
                topInset && { marginTop: topInset },
                bottomInset && { marginBottom: bottomInset },
                bold && boldStyles,
                spacing && { marginHorizontal: spacing },
                style,
            ]}
        />
    );
};

export const verticalDividerStyles = {
    height: '100%',
    width: StyleSheet.hairlineWidth,
    background: 'colors.surfaceVariant',
    bold: {
        width: 1,
    },
};

export default memo(withNormalizedStyleProp(VerticalDivider));
