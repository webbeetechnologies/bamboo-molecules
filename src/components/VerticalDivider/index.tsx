export {
    default as VerticalDivider,
    Props as VerticalDividerProps,
    verticalDividerStyles,
} from './VerticalDivider';
