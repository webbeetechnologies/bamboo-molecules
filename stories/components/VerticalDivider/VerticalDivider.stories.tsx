import type { ComponentMeta, ComponentStory } from '@storybook/react';

import { Example } from './VerticalDivider';

export default {
    title: 'components/VerticalDivider',
    component: Example,
} as ComponentMeta<typeof Example>;

export const Default: ComponentStory<typeof Example> = args => <Example {...args} />;

Default.args = {};

Default.parameters = {
    docs: {
        source: {
            code: `
export const Example = () => {
    const { View, Text } = useMolecules();

    return (
        <View style={styles.container}>
            <Text>Left</Text>
            <VerticalDivider bold spacing={10} />
            <Text>Right</Text>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
    },
});

`,
            language: 'tsx',
            type: 'auto',
        },
    },
};
