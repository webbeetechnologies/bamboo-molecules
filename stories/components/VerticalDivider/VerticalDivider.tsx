import { useMolecules } from '../../../src';
import { VerticalDivider, VerticalDividerProps } from '../../../src/components';
import { StyleSheet } from 'react-native';

export type Props = VerticalDividerProps & {};

export const Example = (props: Props) => {
    const { View, Text } = useMolecules();

    return (
        <View style={styles.container}>
            <Text>Left</Text>
            <VerticalDivider bold spacing={10} {...props} />
            <Text>Right</Text>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
    },
});
