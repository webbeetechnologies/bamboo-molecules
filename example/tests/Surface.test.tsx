import { Example as VerticalDivider } from '../stories/components/VerticalDivider/VerticalDivider';
import { renderWithWrapper } from '../testHelper';

it('should render VerticalDivider', () => {
    const tree = renderWithWrapper(<VerticalDivider />).toJSON();
    expect(tree).toMatchSnapshot();
});

it('should render VerticalDivider with spacing of 10 pixels', () => {
    const tree = renderWithWrapper(<VerticalDivider spacing={10} />).toJSON();
    expect(tree).toMatchSnapshot();
});
