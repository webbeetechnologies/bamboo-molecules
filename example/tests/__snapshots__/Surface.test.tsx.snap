// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render VerticalDivider 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <Text
    style={Object {}}
  >
    Left
  </Text>
  <View
    style={
      Object {
        "background": "rgba(231, 224, 236, 1)",
        "marginHorizontal": 10,
        "width": 1,
      }
    }
  />
  <Text
    style={Object {}}
  >
    Right
  </Text>
</View>
`;

exports[`should render VerticalDivider with spacing of 10 pixels 1`] = `
<View
  style={
    Object {
      "flexDirection": "row",
    }
  }
>
  <Text
    style={Object {}}
  >
    Left
  </Text>
  <View
    style={
      Object {
        "background": "rgba(231, 224, 236, 1)",
        "marginHorizontal": 10,
        "width": 1,
      }
    }
  />
  <Text
    style={Object {}}
  >
    Right
  </Text>
</View>
`;
